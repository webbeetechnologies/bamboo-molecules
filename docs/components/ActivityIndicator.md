# ActivityIndicator

### User Stories
- As a developer, I want to display a spinner in various places to let the user know that there are ongoing activities.
- As a developer, I want to the ability to change the color, and the size of the spinner using props.

### Implementation Details
- Android and IOS will have different styles of spinner.
