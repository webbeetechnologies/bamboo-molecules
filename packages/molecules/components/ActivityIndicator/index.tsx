import { getRegisteredComponentWithFallback, registerMoleculesComponents } from '../../core';
import ActivityIndicatorDefault from './ActivityIndicator';

registerMoleculesComponents({
    ActivityIndicator: ActivityIndicatorDefault,
});

export const ActivityIndicator = getRegisteredComponentWithFallback(
    'ActivityIndicator',
    ActivityIndicatorDefault,
);

export {
    Props as ActivityIndicatorProps,
    activityIndicatorStylesDefault,
    activityIndicatorStyles,
} from './ActivityIndicator';
