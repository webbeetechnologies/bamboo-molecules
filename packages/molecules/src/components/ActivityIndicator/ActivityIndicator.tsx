import { memo, useCallback, useEffect, useMemo, useRef } from 'react';
import {
    Animated,
    Easing,
    Platform,
    View,
    ViewStyle,
    type ActivityIndicatorProps,
} from 'react-native';

import AnimatedSpinner from './AnimatedSpinner';
import { StyleSheet } from 'react-native-unistyles';

export type Props = ActivityIndicatorProps & {
    /**
     * Whether to show the indicator or hide it.
     */
    animating?: boolean;
    /**
     * The color of the spinner.
     */
    color?: string;
    /**
     * Size of the indicator.
     */
    size?: 'small' | 'large' | number;
    /**
     * Whether the indicator should hide when not animating.
     */
    hidesWhenStopped?: boolean;
    style?: ViewStyle;
};

const DURATION = 2400;

const mapIndicatorSize = (indicatorSize: 'small' | 'large' | number | undefined) => {
    if (typeof indicatorSize === 'string') {
        return indicatorSize === 'small' ? 24 : 48;
    }
    return indicatorSize ? indicatorSize : 24;
};

/**
 * Activity indicator is used to present progress of some activity in the app.
 * It can be used as a drop-in for the ActivityIndicator shipped with React Native.
 *
 * <div class="screenshots">
 *   <img src="screenshots/activity-indicator.gif" style="width: 100px;" />
 * </div>
 *
 * ## Usage
 * ```js
 * import * as React from 'react';
 * import { ActivityIndicator, MD2Colors } from 'react-native-paper';
 *
 * const MyComponent = () => (
 *   <ActivityIndicator animating={true} color={MD2Colors.red800} />
 * );
 *
 * export default MyComponent;
 * ```
 */
const ActivityIndicator = ({
    animating = true,
    color: indicatorColorProp,
    hidesWhenStopped = true,
    size: indicatorSize = 'small',
    style: styleProp,
    ...rest
}: Props) => {
    // const componentStyles = useComponentStyles('ActivityIndicator', [
    //     styleProp,
    //     { normalizedIndicatorColorProp: indicatorColorProp },
    // ]);
    const componentStyles = defaultStyles;
    const { current: timer } = useRef<Animated.Value>(new Animated.Value(0));
    const { current: fade } = useRef<Animated.Value>(
        new Animated.Value(!animating && hidesWhenStopped ? 0 : 1),
    );

    const rotation = useRef<Animated.CompositeAnimation | undefined>(undefined);

    const size = mapIndicatorSize(indicatorSize);

    const { color, animationScale, viewStyle, animatedViewStyle } = useMemo(() => {
        const { color: defaultIndicatorColor, animationScale: _animationScale } =
            componentStyles.root;

        return {
            color: indicatorColorProp || defaultIndicatorColor,
            animationScale: _animationScale,
            viewStyle: [componentStyles.container, styleProp],
            animatedViewStyle: [{ width: size, height: size, opacity: fade }],
        };
    }, [
        componentStyles.container,
        componentStyles.root,
        fade,
        indicatorColorProp,
        size,
        styleProp,
    ]);

    const startRotation = useCallback(() => {
        // Show indicator
        Animated.timing(fade, {
            duration: 200 * +animationScale,
            toValue: 1,
            isInteraction: false,
            useNativeDriver: true,
        }).start();

        // Circular animation in loop
        if (rotation.current) {
            timer.setValue(0);
            // $FlowFixMe
            Animated.loop(rotation.current).start();
        }
    }, [animationScale, fade, timer]);

    const stopRotation = () => {
        if (rotation.current) {
            rotation.current.stop();
        }
    };

    useEffect(() => {
        if (rotation.current === undefined) {
            // Circular animation in loop
            rotation.current = Animated.timing(timer, {
                duration: DURATION,
                easing: Easing.linear,
                // Animated.loop does not work if useNativeDriver is true on web
                useNativeDriver: Platform.OS !== 'web',
                toValue: 1,
                isInteraction: false,
            });
        }

        if (animating) {
            startRotation();
        } else if (hidesWhenStopped) {
            // Hide indicator first and then stop rotation
            Animated.timing(fade, {
                duration: 200 * +animationScale,
                toValue: 0,
                useNativeDriver: true,
                isInteraction: false,
            }).start(stopRotation);
        } else {
            stopRotation();
        }

        return () => {
            if (animating) stopRotation();
        };
    }, [animating, fade, hidesWhenStopped, startRotation, animationScale, timer]);

    // console.log({ layer: defaultStyles.layer });

    return (
        <View
            style={viewStyle}
            {...rest}
            accessible
            accessibilityRole="progressbar"
            accessibilityState={{ busy: animating }}>
            <Animated.View style={animatedViewStyle} collapsable={false}>
                {[0, 1].map(index => {
                    // Thanks to https://github.com/n4kz/react-native-indicators for the great work
                    return (
                        <AnimatedSpinner
                            key={index}
                            index={index}
                            size={size}
                            color={color}
                            timer={timer}
                            // style={defaultStyles.layer}
                            duration={DURATION}
                        />
                    );
                })}
            </Animated.View>
        </View>
    );
};

export const defaultStyles = StyleSheet.create(theme => ({
    root: {
        animationScale: theme.animation.scale,
        color: theme.colors.primary,
    } as any,
    container: {
        justifyContent: 'center',
        alignItems: 'center',
    },

    layer: {
        ...StyleSheet.absoluteFillObject,

        justifyContent: 'center',
        alignItems: 'center',
    },
}));

export default memo(ActivityIndicator);
