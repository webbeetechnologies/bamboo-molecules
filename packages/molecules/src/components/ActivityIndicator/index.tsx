import { getRegisteredMoleculesComponent, registerMoleculesComponents } from '../../core';
import ActivityIndicatorDefault from './ActivityIndicator';

registerMoleculesComponents({
    ActivityIndicator: ActivityIndicatorDefault,
});

export const ActivityIndicator = (getRegisteredMoleculesComponent('ActivityIndicator') ??
    ActivityIndicatorDefault) as typeof ActivityIndicatorDefault;

export {
    Props as ActivityIndicatorProps,
    defaultStyles as activityIndicatorStyles,
} from './ActivityIndicator';
