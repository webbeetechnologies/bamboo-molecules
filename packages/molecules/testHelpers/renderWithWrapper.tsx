import type { ReactElement, JSXElementConstructor, ReactNode } from 'react';
import { render, RenderOptions } from '@testing-library/react-native';
import { ProvideMolecules } from '../src';

export const renderWithWrapper = <P extends {} = {}>(
    component: ReactElement<P, string | JSXElementConstructor<P>>,
    options?: RenderOptions,
): ReturnType<typeof render> => {
    const Wrapper = ({ children }: { children: ReactNode }) => (
        <ProvideMolecules>{children}</ProvideMolecules>
    );

    return render(component, { wrapper: Wrapper, ...options });
};
