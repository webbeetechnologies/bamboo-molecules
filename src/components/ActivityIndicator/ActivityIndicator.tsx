import { memo, useCallback, useEffect, useMemo, useRef } from 'react';
import { Animated, Easing, Platform, StyleProp, StyleSheet, ViewStyle } from 'react-native';
import type { ActivityIndicatorProps } from '@webbee/bamboo-atoms';

import type { ComponentStylePropWithVariants } from '../../types';
import { useComponentStyles, useCurrentTheme, useMolecules } from '../../hooks';
import { normalizeStyles } from '../../utils';
import AnimatedSpinner from './AnimatedSpinner';

export type Props = ActivityIndicatorProps & {
    /**
     * Whether to show the indicator or hide it.
     */
    animating?: boolean;
    /**
     * The color of the spinner.
     */
    color?: string;
    /**
     * Size of the indicator.
     */
    size?: 'small' | 'large' | number;
    /**
     * Whether the indicator should hide when not animating.
     */
    hidesWhenStopped?: boolean;
    style?: StyleProp<ViewStyle>;
};

const DURATION = 2400;

const mapIndicatorSize = (indicatorSize: 'small' | 'large' | number | undefined) => {
    if (typeof indicatorSize === 'string') {
        return indicatorSize === 'small' ? 24 : 48;
    }
    return indicatorSize ? indicatorSize : 24;
};

/**
 * Activity indicator is used to present progress of some activity in the app.
 * It can be used as a drop-in for the ActivityIndicator shipped with React Native.
 *
 * <div class="screenshots">
 *   <img src="screenshots/activity-indicator.gif" style="width: 100px;" />
 * </div>
 *
 * ## Usage
 * ```js
 * import * as React from 'react';
 * import { ActivityIndicator, MD2Colors } from 'react-native-paper';
 *
 * const MyComponent = () => (
 *   <ActivityIndicator animating={true} color={MD2Colors.red800} />
 * );
 *
 * export default MyComponent;
 * ```
 */
const ActivityIndicator = ({
    animating = true,
    color: indicatorColorProp,
    hidesWhenStopped = true,
    size: indicatorSize = 'small',
    style: styleProp,
    ...rest
}: Props) => {
    const { View } = useMolecules();
    const theme = useCurrentTheme();
    const { color: indicatorColor, ...style } = useComponentStyles('ActivityIndicator', styleProp);
    const { normalizedIndicatorColorProp } = normalizeStyles(
        { normalizedIndicatorColorProp: indicatorColorProp },
        theme,
    );

    const { current: timer } = useRef<Animated.Value>(new Animated.Value(0));
    const { current: fade } = useRef<Animated.Value>(
        new Animated.Value(!animating && hidesWhenStopped ? 0 : 1),
    );

    const rotation = useRef<Animated.CompositeAnimation | undefined>(undefined);

    const {
        animation: { scale },
    } = theme;

    const color = normalizedIndicatorColorProp || indicatorColor;
    const size = mapIndicatorSize(indicatorSize);

    const viewStyles = useMemo(() => [styles.container, style], [style]);
    const animatedViewStyles = useMemo(
        () => [{ width: size, height: size, opacity: fade }],
        [fade, size],
    );

    const startRotation = useCallback(() => {
        // Show indicator
        Animated.timing(fade, {
            duration: 200 * scale,
            toValue: 1,
            isInteraction: false,
            useNativeDriver: true,
        }).start();

        // Circular animation in loop
        if (rotation.current) {
            timer.setValue(0);
            // $FlowFixMe
            Animated.loop(rotation.current).start();
        }
    }, [scale, fade, timer]);

    const stopRotation = () => {
        if (rotation.current) {
            rotation.current.stop();
        }
    };

    useEffect(() => {
        if (rotation.current === undefined) {
            // Circular animation in loop
            rotation.current = Animated.timing(timer, {
                duration: DURATION,
                easing: Easing.linear,
                // Animated.loop does not work if useNativeDriver is true on web
                useNativeDriver: Platform.OS !== 'web',
                toValue: 1,
                isInteraction: false,
            });
        }

        if (animating) {
            startRotation();
        } else if (hidesWhenStopped) {
            // Hide indicator first and then stop rotation
            Animated.timing(fade, {
                duration: 200 * scale,
                toValue: 0,
                useNativeDriver: true,
                isInteraction: false,
            }).start(stopRotation);
        } else {
            stopRotation();
        }

        return () => {
            if (animating) stopRotation();
        };
    }, [animating, fade, hidesWhenStopped, startRotation, scale, timer]);

    return (
        <View
            style={viewStyles}
            {...rest}
            accessible
            accessibilityRole="progressbar"
            accessibilityState={{ busy: animating }}>
            <Animated.View style={animatedViewStyles} collapsable={false}>
                {[0, 1].map(index => {
                    // Thanks to https://github.com/n4kz/react-native-indicators for the great work
                    return (
                        <AnimatedSpinner
                            key={index}
                            index={index}
                            size={size}
                            color={color}
                            timer={timer}
                            styles={styles}
                            duration={DURATION}
                        />
                    );
                })}
            </Animated.View>
        </View>
    );
};

const styles = {
    container: {
        justifyContent: 'center',
        alignItems: 'center',
    },

    layer: {
        ...StyleSheet.absoluteFillObject,

        justifyContent: 'center',
        alignItems: 'center',
    },
};

export const defaultStyles: ComponentStylePropWithVariants<ViewStyle, '', { color: string }> = {
    color: 'colors.primary',
};

export default memo(ActivityIndicator);
