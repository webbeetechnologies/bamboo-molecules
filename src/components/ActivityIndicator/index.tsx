export {
    default as ActivityIndicator,
    Props as ActivityIndicatorProps,
    defaultStyles as activityIndicatorStyles,
} from './ActivityIndicator';
