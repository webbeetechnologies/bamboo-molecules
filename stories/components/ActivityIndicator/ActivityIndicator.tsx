import { ActivityIndicator, ActivityIndicatorProps } from '../../../src/components';

export type Props = ActivityIndicatorProps & {};

export const Example = (props: Props) => {
    return <ActivityIndicator {...props} />;
};
