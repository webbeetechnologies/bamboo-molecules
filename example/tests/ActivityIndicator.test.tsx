import { Example as ActivityIndicator } from '../stories/components/ActivityIndicator/ActivityIndicator';
import { renderWithWrapper } from '../testHelpers';

it('renders indicator', () => {
    const tree = renderWithWrapper(<ActivityIndicator animating />).toJSON();

    expect(tree).toMatchSnapshot();
});

it('renders hidden indicator', () => {
    const tree = renderWithWrapper(
        <ActivityIndicator animating={false} hidesWhenStopped />,
    ).toJSON();

    expect(tree).toMatchSnapshot();
});

it('renders large indicator', () => {
    const tree = renderWithWrapper(<ActivityIndicator size="large" />).toJSON();

    expect(tree).toMatchSnapshot();
});

it('renders colored indicator', () => {
    const tree = renderWithWrapper(<ActivityIndicator color="#FF0000" />).toJSON();

    expect(tree).toMatchSnapshot();
});
