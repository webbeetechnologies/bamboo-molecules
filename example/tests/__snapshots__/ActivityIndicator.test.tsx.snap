// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders colored indicator 1`] = `
<View
  accessibilityRole="progressbar"
  accessibilityState={
    Object {
      "busy": true,
    }
  }
  accessible={true}
  style={
    Object {
      "alignItems": "center",
      "justifyContent": "center",
      "normalizedIndicatorColorProp": undefined,
    }
  }
>
  <View
    collapsable={false}
    style={
      Object {
        "height": 24,
        "opacity": 1,
        "width": 24,
      }
    }
  >
    <View
      collapsable={false}
      style={
        Object {
          "alignItems": "center",
          "bottom": 0,
          "justifyContent": "center",
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <View
        collapsable={false}
        style={
          Object {
            "height": 24,
            "transform": Array [
              Object {
                "rotate": "45deg",
              },
            ],
            "width": 24,
          }
        }
      >
        <View
          collapsable={false}
          style={
            Object {
              "height": 12,
              "overflow": "hidden",
              "width": 24,
            }
          }
        >
          <View
            collapsable={false}
            style={
              Object {
                "height": 24,
                "transform": Array [
                  Object {
                    "translateY": 0,
                  },
                  Object {
                    "rotate": "-165deg",
                  },
                ],
                "width": 24,
              }
            }
          >
            <View
              collapsable={false}
              style={
                Object {
                  "height": 12,
                  "overflow": "hidden",
                  "width": 24,
                }
              }
            >
              <View
                collapsable={false}
                style={
                  Object {
                    "borderColor": "#FF0000",
                    "borderRadius": 12,
                    "borderWidth": 2.4,
                    "height": 24,
                    "width": 24,
                  }
                }
              />
            </View>
          </View>
        </View>
      </View>
    </View>
    <View
      collapsable={false}
      style={
        Object {
          "alignItems": "center",
          "bottom": 0,
          "justifyContent": "center",
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <View
        collapsable={false}
        style={
          Object {
            "height": 24,
            "transform": Array [
              Object {
                "rotate": "45deg",
              },
            ],
            "width": 24,
          }
        }
      >
        <View
          collapsable={false}
          style={
            Object {
              "height": 12,
              "overflow": "hidden",
              "top": 12,
              "width": 24,
            }
          }
        >
          <View
            collapsable={false}
            style={
              Object {
                "height": 24,
                "transform": Array [
                  Object {
                    "translateY": -12,
                  },
                  Object {
                    "rotate": "345deg",
                  },
                ],
                "width": 24,
              }
            }
          >
            <View
              collapsable={false}
              style={
                Object {
                  "height": 12,
                  "overflow": "hidden",
                  "width": 24,
                }
              }
            >
              <View
                collapsable={false}
                style={
                  Object {
                    "borderColor": "#FF0000",
                    "borderRadius": 12,
                    "borderWidth": 2.4,
                    "height": 24,
                    "width": 24,
                  }
                }
              />
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`renders hidden indicator 1`] = `
<View
  accessibilityRole="progressbar"
  accessibilityState={
    Object {
      "busy": false,
    }
  }
  accessible={true}
  style={
    Object {
      "alignItems": "center",
      "justifyContent": "center",
      "normalizedIndicatorColorProp": undefined,
    }
  }
>
  <View
    collapsable={false}
    style={
      Object {
        "height": 24,
        "opacity": 0,
        "width": 24,
      }
    }
  >
    <View
      collapsable={false}
      style={
        Object {
          "alignItems": "center",
          "bottom": 0,
          "justifyContent": "center",
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <View
        collapsable={false}
        style={
          Object {
            "height": 24,
            "transform": Array [
              Object {
                "rotate": "45deg",
              },
            ],
            "width": 24,
          }
        }
      >
        <View
          collapsable={false}
          style={
            Object {
              "height": 12,
              "overflow": "hidden",
              "width": 24,
            }
          }
        >
          <View
            collapsable={false}
            style={
              Object {
                "height": 24,
                "transform": Array [
                  Object {
                    "translateY": 0,
                  },
                  Object {
                    "rotate": "-165deg",
                  },
                ],
                "width": 24,
              }
            }
          >
            <View
              collapsable={false}
              style={
                Object {
                  "height": 12,
                  "overflow": "hidden",
                  "width": 24,
                }
              }
            >
              <View
                collapsable={false}
                style={
                  Object {
                    "borderColor": "rgba(103, 80, 164, 1)",
                    "borderRadius": 12,
                    "borderWidth": 2.4,
                    "height": 24,
                    "width": 24,
                  }
                }
              />
            </View>
          </View>
        </View>
      </View>
    </View>
    <View
      collapsable={false}
      style={
        Object {
          "alignItems": "center",
          "bottom": 0,
          "justifyContent": "center",
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <View
        collapsable={false}
        style={
          Object {
            "height": 24,
            "transform": Array [
              Object {
                "rotate": "45deg",
              },
            ],
            "width": 24,
          }
        }
      >
        <View
          collapsable={false}
          style={
            Object {
              "height": 12,
              "overflow": "hidden",
              "top": 12,
              "width": 24,
            }
          }
        >
          <View
            collapsable={false}
            style={
              Object {
                "height": 24,
                "transform": Array [
                  Object {
                    "translateY": -12,
                  },
                  Object {
                    "rotate": "345deg",
                  },
                ],
                "width": 24,
              }
            }
          >
            <View
              collapsable={false}
              style={
                Object {
                  "height": 12,
                  "overflow": "hidden",
                  "width": 24,
                }
              }
            >
              <View
                collapsable={false}
                style={
                  Object {
                    "borderColor": "rgba(103, 80, 164, 1)",
                    "borderRadius": 12,
                    "borderWidth": 2.4,
                    "height": 24,
                    "width": 24,
                  }
                }
              />
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`renders indicator 1`] = `
<View
  accessibilityRole="progressbar"
  accessibilityState={
    Object {
      "busy": true,
    }
  }
  accessible={true}
  style={
    Object {
      "alignItems": "center",
      "justifyContent": "center",
      "normalizedIndicatorColorProp": undefined,
    }
  }
>
  <View
    collapsable={false}
    style={
      Object {
        "height": 24,
        "opacity": 1,
        "width": 24,
      }
    }
  >
    <View
      collapsable={false}
      style={
        Object {
          "alignItems": "center",
          "bottom": 0,
          "justifyContent": "center",
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <View
        collapsable={false}
        style={
          Object {
            "height": 24,
            "transform": Array [
              Object {
                "rotate": "45deg",
              },
            ],
            "width": 24,
          }
        }
      >
        <View
          collapsable={false}
          style={
            Object {
              "height": 12,
              "overflow": "hidden",
              "width": 24,
            }
          }
        >
          <View
            collapsable={false}
            style={
              Object {
                "height": 24,
                "transform": Array [
                  Object {
                    "translateY": 0,
                  },
                  Object {
                    "rotate": "-165deg",
                  },
                ],
                "width": 24,
              }
            }
          >
            <View
              collapsable={false}
              style={
                Object {
                  "height": 12,
                  "overflow": "hidden",
                  "width": 24,
                }
              }
            >
              <View
                collapsable={false}
                style={
                  Object {
                    "borderColor": "rgba(103, 80, 164, 1)",
                    "borderRadius": 12,
                    "borderWidth": 2.4,
                    "height": 24,
                    "width": 24,
                  }
                }
              />
            </View>
          </View>
        </View>
      </View>
    </View>
    <View
      collapsable={false}
      style={
        Object {
          "alignItems": "center",
          "bottom": 0,
          "justifyContent": "center",
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <View
        collapsable={false}
        style={
          Object {
            "height": 24,
            "transform": Array [
              Object {
                "rotate": "45deg",
              },
            ],
            "width": 24,
          }
        }
      >
        <View
          collapsable={false}
          style={
            Object {
              "height": 12,
              "overflow": "hidden",
              "top": 12,
              "width": 24,
            }
          }
        >
          <View
            collapsable={false}
            style={
              Object {
                "height": 24,
                "transform": Array [
                  Object {
                    "translateY": -12,
                  },
                  Object {
                    "rotate": "345deg",
                  },
                ],
                "width": 24,
              }
            }
          >
            <View
              collapsable={false}
              style={
                Object {
                  "height": 12,
                  "overflow": "hidden",
                  "width": 24,
                }
              }
            >
              <View
                collapsable={false}
                style={
                  Object {
                    "borderColor": "rgba(103, 80, 164, 1)",
                    "borderRadius": 12,
                    "borderWidth": 2.4,
                    "height": 24,
                    "width": 24,
                  }
                }
              />
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`renders large indicator 1`] = `
<View
  accessibilityRole="progressbar"
  accessibilityState={
    Object {
      "busy": true,
    }
  }
  accessible={true}
  style={
    Object {
      "alignItems": "center",
      "justifyContent": "center",
      "normalizedIndicatorColorProp": undefined,
    }
  }
>
  <View
    collapsable={false}
    style={
      Object {
        "height": 48,
        "opacity": 1,
        "width": 48,
      }
    }
  >
    <View
      collapsable={false}
      style={
        Object {
          "alignItems": "center",
          "bottom": 0,
          "justifyContent": "center",
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <View
        collapsable={false}
        style={
          Object {
            "height": 48,
            "transform": Array [
              Object {
                "rotate": "45deg",
              },
            ],
            "width": 48,
          }
        }
      >
        <View
          collapsable={false}
          style={
            Object {
              "height": 24,
              "overflow": "hidden",
              "width": 48,
            }
          }
        >
          <View
            collapsable={false}
            style={
              Object {
                "height": 48,
                "transform": Array [
                  Object {
                    "translateY": 0,
                  },
                  Object {
                    "rotate": "-165deg",
                  },
                ],
                "width": 48,
              }
            }
          >
            <View
              collapsable={false}
              style={
                Object {
                  "height": 24,
                  "overflow": "hidden",
                  "width": 48,
                }
              }
            >
              <View
                collapsable={false}
                style={
                  Object {
                    "borderColor": "rgba(103, 80, 164, 1)",
                    "borderRadius": 24,
                    "borderWidth": 4.8,
                    "height": 48,
                    "width": 48,
                  }
                }
              />
            </View>
          </View>
        </View>
      </View>
    </View>
    <View
      collapsable={false}
      style={
        Object {
          "alignItems": "center",
          "bottom": 0,
          "justifyContent": "center",
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <View
        collapsable={false}
        style={
          Object {
            "height": 48,
            "transform": Array [
              Object {
                "rotate": "45deg",
              },
            ],
            "width": 48,
          }
        }
      >
        <View
          collapsable={false}
          style={
            Object {
              "height": 24,
              "overflow": "hidden",
              "top": 24,
              "width": 48,
            }
          }
        >
          <View
            collapsable={false}
            style={
              Object {
                "height": 48,
                "transform": Array [
                  Object {
                    "translateY": -24,
                  },
                  Object {
                    "rotate": "345deg",
                  },
                ],
                "width": 48,
              }
            }
          >
            <View
              collapsable={false}
              style={
                Object {
                  "height": 24,
                  "overflow": "hidden",
                  "width": 48,
                }
              }
            >
              <View
                collapsable={false}
                style={
                  Object {
                    "borderColor": "rgba(103, 80, 164, 1)",
                    "borderRadius": 24,
                    "borderWidth": 4.8,
                    "height": 48,
                    "width": 48,
                  }
                }
              />
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
